### PR TITLE
Fixed race condition when cancelling acquire() in Lock or Semaphore

### LINKED ISSUE #387: Potential race condition in `Lock`
Hello!

I'm trying to debug an issue we sometimes hit in production through httpx. We're not using anyio ourselves directly (it's regular asyncio), but httpx is based on it.

I'm not 100% sure the scenario I'm documenting here is the exact issue we're seeing, but there's a good chance. The symptoms we're seeing is that a lock instance gets stuck in the locked state with a finished task still set as the owner.

Here's the rough sketch, involving two tasks:

Task 1:
* acquires the lock

Task 2:
* tries acquiring the lock, fails, creates an event and starts waiting on it

Task 1:
* cancels Task 2, releases the lock setting the lock `_owner_task` to Task 2, gets Task 2's event and sets it

Task 2:
* wakes up with CancelledError on `event.wait()`, propagates it

The lock `_owner_task` is still set to Task 2, rendering the lock stuck.

In our production code it's not Task 1 cancelling Task 2, but aiohttp (but that's not relevant to the example).

Here's a Python snippet demonstrating the problem:
```
from asyncio import CancelledError, create_task, run, sleep
from contextlib import suppress

from anyio import Lock


async def main():
    lock = Lock()

    async def task_b():
        await sleep(0.1)  # Sleep to allow task_a to set up
        async with lock:
            pass

    t = create_task(task_b())

    async def task_a():
        async with lock:
            await sleep(0.1)  # Sleep to allow task_b to try locking
            t.cancel()
        await sleep(0.2)
        print(t.done())  # True
        print(lock._owner_task)  # TaskInfo(id=4337401360, name='Task-2')

    await task_a()
    with suppress(CancelledError):
        await t


run(main())
```
The lock should end up in a free state, but it gets stuck being owned by `task_b`.

### PR DESCRIPTION
**Pull request by agronholm**
Fixes #387.

### CODE REVIEW

File: src/anyio/from_thread.py [hunk 1]
```diff
@@ -82,7 +82,6 @@ async def run_async_cm(self) -> Optional[bool]:
             value = await self._async_cm.__aenter__()
         except BaseException as exc:
             self._enter_future.set_exception(exc)
-            raise
```
Comment (graingert) [inline src/anyio/from_thread.py:85]: Is this related?

Comment (agronholm) [reply src/anyio/from_thread.py:85]: Whoa, how did that get into the PR...must've been an artifact of my work on that other PR. Good job noticing this!

File: src/anyio/_core/_synchronization.py [hunk 2]
```diff
@@ -302,6 +304,8 @@ async def acquire(self) -> None:
             except BaseException:
                 if not event.is_set():
                     self._waiters.remove(event)
+                else:
+                    self._value += 1
```
Comment (agronholm) [inline src/anyio/_core/_synchronization.py:308]: I'm thinking, maybe this should call `self.release()` too, like the `Lock` version?

**Review by graingert — APPROVED** [review]

**Comment by agronholm** [comment]
Thanks for the review!
